programs/clearing_house/src/lib.rs

```diff
@@ -754,6 +758,234 @@ pub mod clearing_house {
+
+        if lp_tokens_to_burn == 0 {
+            return Ok(());
+        }
+
+        validate!(
+            lp_position.lp_tokens >= lp_tokens_to_burn,
+            ErrorCode::InsufficientLPTokens,
+            "Trying to burn more lp tokens than the user has",
+        )?;
+
+        // settle the lp first
+        let settle_result = settle_lp_position(lp_position, lp_tokens_to_burn, &mut market)?;
+
+        validate!(
+            settle_result != SettleResult::DidNotRecieveMarketPosition,
+            ErrorCode::UnableToBurnLPTokens
+        )?;
+
+        // transform lp_position into a market position
+        lp_position.lp_tokens = lp_position
+            .lp_tokens
+            .checked_sub(lp_tokens_to_burn)
+            .ok_or_else(math_error!())?; // burn tokens
+
+        // update funding rate to the markets position
+        lp_position.last_cumulative_funding_rate = if lp_position.base_asset_amount > 0 {
+            market.amm.cumulative_funding_rate_long
+        } else {
+            market.amm.cumulative_funding_rate_short
+        };
+
+        // track new position in the market
+        // TODO: probably want to refactor all this into one fcn (auction PR fcn?)
+        market.amm.net_base_asset_amount = market
+            .amm
+            .net_base_asset_amount
+            .checked_add(lp_position.base_asset_amount)
+            .ok_or_else(math_error!())?;
+
+        let base_asset_acquired = lp_position.base_asset_amount;
+        let quote_asset_amount = lp_position.quote_asset_amount;
+
+        if lp_position.base_asset_amount > 0 {
+            market.base_asset_amount_long = market
+                .base_asset_amount_long
+                .checked_add(base_asset_acquired)
+                .ok_or_else(math_error!())?;
+            market.amm.quote_asset_amount_long = market
+                .amm
+                .quote_asset_amount_long
+                .checked_add(quote_asset_amount)
+                .ok_or_else(math_error!())?;
+        } else {
+            market.base_asset_amount_short = market
+                .base_asset_amount_short
+                .checked_add(base_asset_acquired)
+                .ok_or_else(math_error!())?;
+            market.amm.quote_asset_amount_short = market
+                .amm
+                .quote_asset_amount_short
+                .checked_add(quote_asset_amount)
+                .ok_or_else(math_error!())?;
+        }
+
+        market.open_interest = market
+            .open_interest
+            .checked_add(1)
+            .ok_or_else(math_error!())?;
+
+        // update market state
+        let new_sqrt_k = market
+            .amm
+            .sqrt_k
+            .checked_sub(lp_tokens_to_burn)
+            .ok_or_else(math_error!())?;
+        let new_sqrt_k_u192 = bn::U192::from(new_sqrt_k);
+
+        let update_k_result = get_update_k_result(&market, new_sqrt_k_u192)?;
+        math::amm::update_k(&mut market, &update_k_result)?;
```

Comment:
probably want to not throw an error on this k update if its a decrease >2.2%, since its all or nothing remove liq rn?